zerver/views/custom_profile_fields.py

```diff
@@ -0,0 +1,93 @@
+from django.utils.translation import ugettext as _
+
+from zerver.decorator import has_request_variables, REQ, require_realm_admin
+from zerver.forms import ProfileFieldForm, ProfileFieldValueForm
+from zerver.lib.actions import (do_add_realm_custom_profile_field,
+                                do_remove_realm_custom_profile_field,
+                                do_update_realm_custom_profile_field,
+                                do_update_user_custom_profile_data)
+from zerver.lib.response import json_success, json_error
+from zerver.lib.validator import check_dict, check_list, check_int, check_string
+
+from zerver.models import (custom_profile_fields_for_realm, UserProfile,
+                           CustomProfileField, custom_profile_fields_for_realm)
+
+def list_realm_custom_profile_fields(request, user_profile):
+    # type: (HttpRequest, UserProfile) -> HttpResponse
+    fields = custom_profile_fields_for_realm(user_profile.realm_id)
+    return json_success({'custom_fields': [f.as_dict() for f in fields]})
+
+@require_realm_admin
+@has_request_variables
+def create_realm_custom_profile_field(request, user_profile, name=REQ(),
+                                      field_type=REQ(validator=check_int)):
+    # type: (HttpRequest, UserProfile, Text, int) -> HttpResponse
+    form = ProfileFieldForm(data={'name': name, 'field_type': field_type})
+    if not form.is_valid():
+        return json_error(_("Invalid data supplied."), data=form.errors)
+
+    try:
+        field = do_add_realm_custom_profile_field(
+            realm=user_profile.realm,
+            name=name,
+            field_type=field_type,
+        )
+        return json_success({'id': field.id})
+    except IntegrityError:
+        return json_error(_("Field name should be unique in an "
+                            "organization."))
+
+@require_realm_admin
+def delete_realm_custom_profile_field(request, user_profile, field_id):
+    # type: (HttpRequest, UserProfile, int) -> HttpResponse
+    try:
+        field = CustomProfileField.objects.get(id=field_id)
+    except CustomProfileField.DoesNotExist:
+        return json_error(_('Field not found'))
+
+    do_remove_realm_custom_profile_field(realm=user_profile.realm,
+                                         field=field)
+    return json_success()
+
+@require_realm_admin
+@has_request_variables
+def update_realm_custom_profile_field(request, user_profile, field_id,
+                                      name=REQ()):
+    # type: (HttpRequest, UserProfile, int, Text) -> HttpResponse
+    realm = user_profile.realm
+    try:
+        field = CustomProfileField.objects.get(realm=realm, id=field_id)
+    except CustomProfileField.DoesNotExist:
+        return json_error(_('Field not found.'))
+
+    try:
+        do_update_realm_custom_profile_field(realm, field, name)
+    except IntegrityError:
+        return json_error(_('Name of the field should be unique in a realm.'))
```

Comment:
This error message, and the similar one above, should probably be "A field with that name already exists!".

API users shouldn't need to be aware of whether other realms/organizations exist, so no need to mention that in error messages.